### PR TITLE
Add OS credential-store password fallbacks

### LINKED ISSUE #886: Secret resolver (OS keychain)
## Background
You may need to handle secret values ​​as well as config and env.

Secret values ​​must not be used in config, and must be received only from the OS keychain or env.

## API
The following is an example where the keyring acts as a fallback when an environment exists, such as in CI or tests.
Assume there is a hypothetical function called `keyring`.
There might be `createKeyringContext`, just like config and env.
```typescript
const envContext = createEnvContext({ prefix: "MYAPP_" });
const parser = object({
  apiToken: map(
    seq(
      optional(bindEnv(fail<string>(), { context: envContext, key: "API_TOKEN" })),
      optinoal(keyring({ service: "files", username: "default" }))
    ),
    ([env, keyring]) => env ?? keyring
  )
})
```

## Reference
Here is an example of a package that can use the OS keychain.

- https://github.com/Brooooooklyn/keyring-node
- https://github.com/biw/keychain-store

For reference, while keyring-node supports multiple platforms, it has the disadvantage of not being able to use biometric authentication(a.k.a Touch ID) on a Mac.

To use biometric authentication, you need `apple_native_keyring_store::proctected` instead of `apple_native_keyring_store::keychain`

- https://github.com/open-source-cooperative/apple-native-keyring-store
- https://github.com/Brooooooklyn/keyring-node/blob/a8709c1381e604c22b9a01b78277a51d65447a69/src/entry.rs#L33
```rust
fn setup_macos_store() -> anyhow::Result<()> {
  use std::collections::HashMap;

  use apple_native_keyring_store::keychain::Store;

  let store = Store::new_with_configuration(&HashMap::new())?;
  keyring_core::set_default_store(store);
  Ok(())
}
```

### PR DESCRIPTION
**Pull request by black7375**
Applications can use OS credential-store passwords as parser fallbacks without
storing secrets in configuration files. Add *@optique/keyring* with
`createKeyringContext()` and `bindKeyring()`, using `@napi-rs/keyring`'s
asynchronous `AsyncEntry` backend and an injectable source for tests or
alternative backends.

Explicit CLI values and outer sources such as `bindEnv()` take precedence.
The native addon loads only when a keyring fallback is reached; help, version,
and suggestions bypass it. Missing credentials allow the inner fallback, while
store errors propagate. Retrieved passwords pass through inner validation,
with validation failures redacted. Lookups are cached within each parser
occurrence of a run without retaining credentials across runs.

Includes npm/JSR package integration and documentation for hidden password
prompts, environment overrides, context registration, and Deno permissions.

Tests pass across Deno, Node.js, and Bun, including native-loading smoke tests.
Type, lint, formatting, publish dry-run, and changelog checks pass. All 741
documentation examples pass Twoslash checks; full site bundling remains
unverified because the local build exceeded its time limit.

Fixes https://github.com/dahlia/optique/issues/886.

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -182,6 +182,22 @@ To be released.
...
+    applications can keep secrets out of configuration files. Missing
+    credentials allow a fallback, while locked, inaccessible, or ambiguous
+    credential-store errors reject the parse. Keyring values take precedence
+    over inner fallbacks, including when used as dependency sources. The
+    credential store is accessed only when a password fallback is needed,
+    including in two-pass runs. Bound parsers preserve the inner parser's
+    help output and dependency sources inside selected commands and
+    alternatives. Stored-password validation failures hide their original
+    details so passwords do not appear in error output.
+    [[#886] by black7375\]
```
[the remark this answers is not in the document]
Comment (dahlia) [reply CHANGES.md]: It's automatically formatted by Sacho. Never mind.

File: packages/keyring/src/native.test.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      "--allow-read=../../node_modules",
+      "--allow-ffi=../../node_modules",
+      fixturePath,
+    ]
+    : [fixturePath];
+  const result = spawnSync(process.execPath, args, {
+    cwd: packagePath,
+    encoding: "utf8",
+    env: environment,
+  });
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/keyring/src/native.test.ts:69]: I'd keep the fixture environment isolated. Inheriting all of `process.env` could let settings such as `NAPI_RS_NATIVE_LIBRARY_PATH` affect the normal-loading test. These tests currently pass under Deno, Node.js, and Bun on Linux.

If a platform needs specific variables, we can pass those explicitly once we have a failing case. I'm resolving this for now, since the review doesn't identify a required variable or a reproducible failure.

File: packages/keyring/src/parser.ts [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+      const result = await withInnerState(
+        context.state,
+        (annotatedState) => parser.parse({ ...context, state: annotatedState }),
+      );
+      if (!result.success && result.consumed > 0) return result;
+      const nextState = injectAnnotations({
+        [stateKey]: stateId,
+        hasCliValue:
+          (isBindState(context.state) && context.state.hasCliValue) ||
+          (result.success && result.consumed.length > 0),
```
Comment (dahlia) [inline packages/keyring/src/parser.ts]: Could we exclude consumption of the option terminator from `hasCliValue`? `option()` returns success after consuming `--`, but leaves its value state unchanged. This condition treats that as an explicit CLI value, so `complete()` skips the keyring.

With a registered source returning `"stored-password"`, `bindKeyring(option("--password", string()), options)` succeeds for `[]` but fails with `Missing option --password.` for `["--"]`, without calling the source. I reproduced this under Deno, Node.js, and Bun on Linux.

Please preserve the fallback when only the terminator was consumed and add a regression test for this case.

File: packages/keyring/src/parser.ts [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+  );
+  const schedulingNodes = boundParser[effectfulSchedulingNodesKey];
+  if (schedulingNodes != null) {
+    Object.defineProperty(boundParser, effectfulSchedulingNodesKey, {
+      // Only CLI-selected inner parsers may contribute sources before this
+      // binding tries its keyring fallback. Keep that policy local to keyring.
+      value: ((state, path) =>
+        isBindState(state) && state.hasCliValue
+          ? schedulingNodes(state, path)
+          : []) satisfies EffectfulSchedulingNodesFn,
```
Comment (dahlia) [inline packages/keyring/src/parser.ts]: Could we preserve inner dependency scheduling when the outer keyring returns `undefined`? Hiding these nodes before trying the outer source preserves precedence, but when the outer source returns `undefined`, the inner fallback's dependency value does not reach a sibling consumer.

I reproduced this with the following composition, where `innerContext` returns `"prod"` and `outerContext` returns `undefined` (both contexts are registered):

~~~~ typescript
const mode = dependency(choice(["dev", "prod"] as const));
const level = mode.derive({
  metavar: "LEVEL",
  mode: "sync",
  factory: (value: "dev" | "prod") =>
    choice(value === "dev" ? ["debug"] as const : ["silent"] as const),
  defaultValue: () => "dev" as const,
});
const inner = object({
  mode: bindKeyring(option("--mode", mode), {
    context: innerContext, service: "test", username: "test",
  }),
}).map(value => value.mode);
const parser = object({
  mode: bindKeyring(inner, {
    context: outerContext, service: "test", username: "test",
  }),
  level: option("--level", level),
});
~~~~

`--level silent` is rejected against the default `"dev"` branch, even though the inner source supplies `"prod"`. Conversely, `--level debug` can succeed with `{ mode: "prod", level: "debug" }`. Replacing the outer binding with `inner` restores the expected validation. I reproduced the rejection of `--level silent` under Deno, Node.js, and Bun on Linux.

Please add coverage for an absent outer source around a mapped construct with an inner dependency source, including rejection of values that are invalid for the resolved source.

File: docs/integrations/keyring.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+Missing credentials and store errors
+------------------------------------
+
+When no credential exists, `bindKeyring()` lets the inner parser supply its
+fallback or missing-value error. `optional(bindKeyring(...))` returns
+`undefined` for a missing credential, and `withDefault(bindKeyring(...), value)`
+uses the default only when the keyring cannot supply a value.
+
+Locked, inaccessible, or ambiguous credential-store errors reject the parse
+with the original error. They do not trigger an inner fallback or get replaced
```
Comment (dahlia) [inline docs/integrations/keyring.md]: Could we clarify how this guarantee applies to Linux store initialization? In `@napi-rs/keyring` 2.0.0, [`LinuxCredentialBuilder::new()`](https://github.com/Brooooooklyn/keyring-node/blob/v2.0.0/src/linux_credential_builder.rs#L15-L18) discards any Secret Service initialization error and falls back to the kernel keyring. The original initialization error therefore never reaches this wrapper; if the kernel keyring has no matching entry, the initialization failure could appear to be an ordinary missing credential.

This differs from the lookup-result handling, which correctly propagates errors other than `NoEntry`. I confirmed the automatic store switch in the backend source, but have not reproduced the resulting lookup behavior against a live store.

The distinction between a missing entry and an inaccessible store was one of the requirements in issue #886. Can we establish whether this backend behavior meets that requirement, and make the implementation and documented guarantee agree? The native smoke tests currently check loading and deferred loading failure, so they do not establish this behavior.

File: changes.d/keyring/password-source.md [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+    applications can keep secrets out of configuration files. Missing
+    credentials allow a fallback, while locked, inaccessible, or ambiguous
+    credential-store errors reject the parse. Keyring values take precedence
+    over inner fallbacks, including when used as dependency sources. The
+    credential store is accessed only when a password fallback is needed,
+    including in two-pass runs. Bound parsers preserve the inner parser's
+    help output and dependency sources inside selected commands and
+    alternatives. Stored-password validation failures hide their original
+    details so passwords do not appear in error output.
+    [[#886] by black7375]
```
Comment (dahlia) [inline changes.d/keyring/password-source.md]: Please include this PR reference as well as the issue reference in the fragment: `[[#886], [#956] by Your Name]`.

For contributor credit, I prefer your name rather than your GitHub handle, unless you specifically want to be credited under a pseudonym. Please use the name you would like published; keeping `black7375` is fine if that is your preferred pseudonym.

After updating the fragment, please regenerate the materialized entry with `sacho sync`.

**Review by dahlia — CHANGES_REQUESTED** [review]
Thanks for putting this together and for the thorough test coverage. I've left a few inline comments on parser edge cases, the Linux backend's error behavior, and the changelog.

File: packages/keyring/src/parser.ts (hunk 3 above)
Comment (black7375) [reply packages/keyring/src/parser.ts]: Fixed in 93582cca. Consuming only the option terminator no longer sets `hasCliValue`, so the keyring fallback remains available. Added regression tests for synchronous and asynchronous inner parsers, plus a test preserving a literal `--` supplied as a positional value.

File: packages/keyring/src/parser.ts (hunk 4 above)
Comment (black7375) [reply packages/keyring/src/parser.ts]: Fixed in b7761dd8. When the outer keyring returns undefined, the inner fallback is now prepared before sibling dependency consumers are validated. The prepared result is reused during final completion.

Added coverage for both field orders: resolving `prod` accepts `silent` and rejects `debug`. The tests also cover demanded and undemanded inner sources during two-pass runs.

File: docs/integrations/keyring.md (hunk 5 above)
Comment (black7375) [reply docs/integrations/keyring.md]: Addressed in 2606a7d5. You're right that the native backend's initialization fallback could hide the original error.

The Linux default source now accesses Secret Service directly over D-Bus using encrypted sessions. It returns undefined only after a successful search with no matching credential.
Locked or ambiguous matches and connection, session, or lookup failures reject without switching stores. Other platforms continue to use AsyncEntry.

File: changes.d/keyring/password-source.md (hunk 6 above)
Comment (black7375) [reply changes.d/keyring/password-source.md]: Updated in `692df600`. The fragment now includes both the issue and PR references, with `Minseok Youn` as my preferred published name.

**Review by dahlia — CHANGES_REQUESTED** [review]
Thanks for addressing the review comments! One small correction remains: all six commits use `Assisted-by: Codex:gpt-6`. Could you confirm which model you used and update the trailers accordingly? If it was GPT-6 Astra, the [official Model ID](https://developers.openai.com/api/docs/models/gpt-6-astra) is `gpt-6-astra`, so the trailer should be `Assisted-by: Codex:gpt-6-astra`.

**Comment by black7375** [comment]
I have now updated the commit message and force-pushed it.

**Review by dahlia — APPROVED** [review]
Thanks for working on it! Great job.

**Comment by black7375** [comment]
Thank you for the helpful guide.
The best CLI parser. I'm making good use of it.
